### ISSUE TITLE
deprecate isscalar, change behavior?

### ISSUE THREAD

**Issue by nschloe** [numpy/numpy#19690]
Right now, `np.isscalar` returns `True` iff the input is a Python scalar:
```python
import numpy as np

np.isscalar(3.14)              # True
np.isscalar(np.array(3.14))    # False
np.isscalar(np.array([3.14]))  # False
```
However, `np.array`s with `ndim == 0` smell, taste, and behave like Python scalars in almost all situations, so you almost always want to admit ndim-0 arrays when writing `if np.isscalar(a): ...`. Unfortunately, downstream uses `isscalar` far more often (e.g. scipy: 61 `isscalar`, 9 `ndim`). This begs the question if `np.isscalar` should be deprecated or its behavior changed to
```python
def isscalar(x):
    return np.ndim(x) == 0
```

**Comment by eric-wieser** [comment]
#11882 was the first step towards a possible deprecation, which documented the reasons to avoid `isscalar`.

**Comment by pankaj-cipher** [comment]
Talking about the above situation depreciating `np.isscalar()` . I would rather suggest deprecating 0D-arrays because 0D-arrays are no more arrays , they are scalars and obviously a `array(vector)` couldn't be a `scalar` . So, basically rather changing the definition for `np.isscalar()`, it should be changed for `np.array(scalar)`.

**Comment by nschloe** [comment]
> 0D-arrays are no more arrays , they are scalars

I agree! I wouldn't vote for their deprecation though because there are many situations where you want scalars to be treated just like arrays, e.g., when evaluating functions ("vectorization").

**Comment by eric-wieser** [comment]
I don't think changing the behavior of `isscalar`this late in the game is sensible. Even numpy itself depends on the weirdness of its behavior (it uses it to mean "can I modify this safely in place with `+=`" in some instances); so I would expect many downstream breakages. Yes, we should have defined it that way in the first place, but given we didn't, and that `np.ndim(x) == 0` is only one character longer, it's not worth the breakage.

I'd argue that `isscalar` isn't even a better name; there's no clear definition of what counts as a `scalar`, so users read it as whichever meaning they think they want at the moment.

0D arrays have their place, and I don't think for a moment would be a good idea to remove

**Comment by seberg** [comment]
> 0D arrays have their place, and I don't think for a moment would be a good idea to remove

It is a common argument to get rid of scalars, but never getting rid of 0-D arrays :).

What is a "scalar" in NumPy depends, there are at least three definitions:
1. A scalar is equivalent to `np.asarray(x, dtype=dtype).ndim == 0`
2. Same as 1., but do not accept anything deemed array-like
3. Same as 2., but only accept "recognized" scalars (i.e. an arbitrary object is not a scalar, we never probe whether it is an array-like).  An example would also be `class o: def __float__(self): 1.`, which can be stored in a float array, but is not recognized. In this definition even a subclass of a float may be rejected.

(I guess 2. and 3. could also be solved with a three-way function that returns `scalar`, `array-like`, `unknown`, and possibly `array-like-scalar`)

And note that the `dtype=dtype` actually has an effect.  As an extreme case `dtype=object` means that _anything_ is technically a scalar (even a 0-D array).  A tuple is a scalar if the input is a structured dtype.  A vector object could be an array-like unless the `dtype=vector`.

---

We should not change `isscalar` to mean `ndim == 0` IMO.  We might be able to tweak it to be better defined (I am not actually sure what it roughly does currently) with respect to 2. or 3.
If we do not need 2. or 3., then I would agree that deprecating is probably just as well.  Even if we need 2. or 3. it seems likely better to just add something better defined and deprecate it anyway (rather than trying to fix `isscalar`).
